pyro/infer/csis.py

```diff
@@ -0,0 +1,161 @@
+from __future__ import absolute_import, division, print_function
+
+import pyro
+import pyro.poutine as poutine
+from pyro.infer.importance import Importance
+from pyro.infer.util import torch_backward, torch_item
+from pyro.poutine.util import prune_subsample_sites
+from pyro.util import check_model_guide_match, warn_if_nan
+
+
+class CSIS(Importance):
+    """
+    Compiled Sequential Importance Sampling, allowing compilation of a guide
+    program to minimise KL(model posterior || guide), and inference with
+    importance sampling.
+
+    **Reference**
+    "Inference Compilation and Universal Probabilistic Programming" `pdf https://arxiv.org/pdf/1610.09900.pdf`
+
+    :param model: probabilistic model defined as a function. Must accept
+        keyword arguments with names of observed sites in model.
+    :param guide: guide function which is used as an approximate posterior. Must
+        accept keyword arguments with names of observed sites in model.
```

Comment:
nit: instead of one keyword argument per site, what about requiring the guide to take a single keyword argument `observations`, a dictionary whose keys are names of observed sites?  That feels like a more idiomatic way to write guides that take lots of observed sites with mechanically generated names, e.g. time series models.

Reply:
That's a good idea, I'll change that